### PR TITLE
Emit counter metrics tracking nftables set usage

### PR DESCRIPTION
**Pull request by rkolavo**
This evaluates in the set_manager how much each element of a set has been seen since it last reset and emits a counter for that value in each packets and bytes. When applicable, the Address/Address Range/Prefix or Port/Port Range is used as a tag on the counter.

### CODE REVIEW

File: pkg/set/set_manager.go [hunk 1]
```diff
@@ -139,3 +147,32 @@ func (s *ManagedSet) genTags(additional []string) []string {
...
+		} else if err := utils.ValidatePrefix(d.setData.Prefix); err != nil {
+			tags = []string{fmt.Sprintf("startip_endip:%s", d.setData.Prefix)}
+		} else if err := utils.ValidateAddress(d.setData.Address); err != nil {
+			tags = []string{fmt.Sprintf("startip_endip:%s", d.setData.Address)}
+		} else if err := utils.ValidateAddressRange(d.setData.AddressRangeStart, d.setData.AddressRangeEnd); err != nil {
+			tags = []string{fmt.Sprintf("startip_endip:%s-%s", d.setData.AddressRangeStart, d.setData.AddressRangeEnd)}
+		} else {
+			s.logger.Warnf("invalid set data encountered while emitting counter metrics: %+v", d.setData)
+			continue
+		}
```
Comment (rkolavo) [inline pkg/set/set_manager.go:167]: I'm open to adding validation methods directly to SetData like `hasValidPortRange()`, `hasValidPrefix()` to perhaps make this a bit more readable

File: pkg/set/set.go [hunk 2]
```diff
@@ -188,6 +188,84 @@ func (s *Set) GetSet() *nftables.Set {
...
 }
 
+// Gets a set's SetData with associated counters
+func (s *Set) getCountedSetData(c *nftables.Conn) ([]countedSetData, error) {
+	elements, err := c.GetSetElements(s.set)
+	if err != nil {
+		return nil, err
+	}
+
+	if s.set.KeyType == nftables.TypeIPAddr || s.set.KeyType == nftables.TypeIP6Addr {
```
Comment (joewilliams) [inline pkg/set/set.go:198]: Please use a `switch` for this similar to elsewhere in this module.

File: pkg/set/set.go [hunk 3]
```diff
@@ -188,6 +188,84 @@ func (s *Set) GetSet() *nftables.Set {
...
+	return nil, fmt.Errorf("unexpected set key type: %v", s.set.KeyType)
+}
+
+func countedPortSetData(elements []nftables.SetElement) ([]countedSetData, error) {
+	setDataList := []countedSetData{}
+
+	// set elements come in pairs, first the end of range, then start of range which contains counters
+	for i := 0; i < len(elements); i++ {
+		endElement := elements[i]
+		if !endElement.IntervalEnd {
```
Comment (joewilliams) [inline pkg/set/set.go:213]: Maybe we can create a function that will return the start and end elements so we can use the same code in both `countedPortSetData` and `countedAddrSetData`?

File: pkg/set/set_data.go [hunk 4]
```diff
@@ -260,3 +269,49 @@ func NetipAddrPortsToSetData(addrports []netip.AddrPort) ([]SetData, []SetData,
 func NetipAddrPortToSetData(addrport netip.AddrPort) (SetData, SetData, error) {
 	return SetData{Address: addrport.Addr()}, SetData{Port: uint16(addrport.Port())}, nil
 }
+
+// Convert start and end address bytes to SetData type
+func AddressBytesToSetData(start []byte, end []byte) (SetData, error) {
```
Comment (joewilliams) [inline pkg/set/set_data.go:274]: Please make sure we have 100% test coverage for the various branches in this function and `PortBytesToSetData`.

**Review by joewilliams — CHANGES_REQUESTED** [review]

File: pkg/set/set_manager.go (hunk 1 above)
Comment (blainsmith) [reply pkg/set/set_manager.go:167]: To address the readability you can also do:

```go
switch {
case utils.ValidatePort(d.setData.Port) != nil:
  tags = []string{fmt.Sprintf("startip_endip:%v", d.setData.Port)}

case utils.ValidatePortRange(d.setData.PortRangeStart, d.setData.PortRangeEnd) != nil:
  tags = []string{fmt.Sprintf("startip_endip:%v-%v", d.setData.PortRangeStart, d.setData.PortRangeEnd)}
...

default:
  s.logger.Warnf("invalid set data encountered while emitting counter metrics: %+v", d.setData)
  continue
}
```

This approach removes a lot of the `if else` noise which is a pretty common patter for lots of `if else` branches with more then 2.

**Review by blainsmith — CHANGES_REQUESTED** [review]

File: pkg/set/set_manager.go (hunk 1 above)
Comment (rkolavo) [reply pkg/set/set_manager.go]: thank you 🙏🏻

**Review by joewilliams — APPROVED** [review]
Looking good to me, thank you!

**Review by blainsmith — APPROVED** [review]

**Review by masonjjohnson — APPROVED** [review]

**Pull request merged**